README.md
![CompilerGym](https://github.com/facebookresearch/CompilerGym/raw/development/docs/source/_static/img/logo-padded.png)

<p align="center">
  <!-- Getting started colab -->
  <a href="https://colab.research.google.com/github/facebookresearch/CompilerGym/blob/stable/examples/getting-started.ipynb">
      <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Colab" height="20">
  </a>
  <!-- Supported python versions list -->
  <a href="https://pypi.org/project/compiler-gym/">
      <img src="https://img.shields.io/pypi/pyversions/compiler-gym" alt="Python versions" height="20">
  </a>
  <!-- Downloads counter -->
  <a href="https://pypi.org/project/compiler-gym/">
      <img src="https://pepy.tech/badge/compiler-gym" alt="PyPi Downloads" height="20">
  </a>
  <!-- PyPi Version -->
  <a href="https://pypi.org/project/compiler-gym/">
      <img src="https://badge.fury.io/py/compiler-gym.svg" alt="PyPI version" height="20">
  </a>
  <!-- license -->
  <a href="https://tldrlegal.com/license/mit-license">
      <img src="https://img.shields.io/pypi/l/compiler-gym" alt="License" height="20">
  </a>
  <!-- CI status -->
  <a href="https://github.com/facebookresearch/CompilerGym/actions?query=workflow%3ACI+branch%3Adevelopment">
      <img src="https://github.com/facebookresearch/CompilerGym/workflows/CI/badge.svg?branch=development" alt="CI status" height="20">
  </a>
</p>

<p align="center">
  <i>Reinforcement learning environments for compiler optimization tasks.</i>
</p>
<p align="center">
  <i>
    Check
    <a href="http://facebookresearch.github.io/CompilerGym/">the website</a>
    for more information.
  </i>
</p>


## Introduction

CompilerGym is a library of easy to use and performant reinforcement learning
environments for compiler tasks. It allows ML researchers to interact with
important compiler optimization problems in a language and vocabulary with which
they are comfortable, and provides a toolkit for systems developers to expose
new compiler tasks for ML research. We aim to act as a catalyst for making
compilers faster using ML. Key features include:

* **Ease of use:** built on the the popular [Gym](https://gym.openai.com/)
  interface - use Python to write your agent. With CompilerGym, building ML
  models for compiler research problems is as easy as building ML models to play
  video games.

* **Batteries included:** includes everything required to get started. Wraps
  real world programs and compilers to provide millions of instances for
  training. Provides multiple kinds of pre-computed program representations: you
  can focus on end-to-end deep learning or features + boosted trees, all the way
  up to graph models. Appropriate reward functions and loss functions for
  optimization targets are provided out of the box.

* **Reproducible:** provides validation for correctness of results, common
  baselines, and [leaderboards](#leaderboards) for you to submit your results.

For a glimpse of what's to come, check out [our
roadmap](https://github.com/facebookresearch/CompilerGym/projects/1).


## News

- November 2022: CompilerGym v0.2.5 adds support for Python 3.10 and drops
  support for 3.7. See [release
  notes](https://github.com/facebookresearch/CompilerGym/releases/tag/v0.2.5)
  for full details.
- April 2022: ⭐️ CompilerGym wins the **distinguished paper award** at
  [CGO'22](https://conf.researchr.org/home/cgo-2022)! You can read our work
  [here](https://arxiv.org/pdf/2109.08267.pdf).
- April 2022: 📖 Our **tutorial** at
  [CGO'22](https://conf.researchr.org/home/cgo-2022) was well attended. If you
  missed the event, you can work through the materials
  [here](https://chriscummins.cc/2022/cgo22-tutorial/).
- September 2021: 📄 CompilerGym was featured on the **Meta AI research blog**.
  You can read the post
  [here](https://ai.facebook.com/blog/compilergym-making-compiler-optimizations-accessible-to-all/).


## Installation

Install the latest CompilerGym release using:

    pip install -U compiler_gym

See
[INSTALL.md](https://github.com/facebookresearch/CompilerGym/blob/development/INSTALL.md)
for further details.


## Usage

Starting with CompilerGym is simple. If you not already familiar with the gym
interface, refer to the [getting started
guide](http://facebookresearch.github.io/CompilerGym/getting_started.html) for
an overview of the key concepts.

In Python, import `compiler_gym` to use the environments:

```py
>>> import compiler_gym                      # imports the CompilerGym environments
>>> env = compiler_gym.make(                 # creates a new environment (same as gym.make)
...     "llvm-v0",                           # selects the compiler to use
...     benchmark="cbench-v1/qsort",         # selects the program to compile
...     observation_space="Autophase",       # selects the observation space
...     reward_space="IrInstructionCountOz", # selects the optimization target
... )
>>> env.reset()                              # starts a new compilation session
>>> env.render()                             # prints the IR of the program
>>> env.step(env.action_space.sample())      # applies a random optimization, updates state/reward/actions
>>> env.close()                              # closes the environment, freeing resources
```

See the [examples](/examples) directory for agent implementations, environment
extensions, and more. See the [documentation
website](http://facebookresearch.github.io/CompilerGym/) for the API reference.


## Leaderboards

These leaderboards track the performance of user-submitted algorithms for
CompilerGym tasks. To submit a result please see
[this document](https://github.com/facebookresearch/CompilerGym/blob/development/CONTRIBUTING.md#leaderboard-submissions).


### LLVM Instruction Count

LLVM is a popular open source compiler used widely in industry and research. The
`llvm-ic-v0` environment exposes LLVM's optimizing passes as a set of actions
that can be applied to a particular program. The goal of the agent is to select
the sequence of optimizations that lead to the greatest reduction in instruction
count in the program being compiled. Reward is the reduction in instruction
count achieved scaled to the reduction achieved by LLVM's builtin `-Oz`
pipeline.

This leaderboard tracks the results achieved by algorithms on the `llvm-ic-v0`
environment on the 23 benchmarks in the `cbench-v1` dataset.

| Author                                           | Algorithm | Links | Date | Walltime (mean) | Codesize Reduction (geomean) |
|--------------------------------------------------| --- | --- | --- |-----------------|------------------------------|
 | Robin Schmöcker, Yannik Mahlau, Nicolas Fröhlich | PPO + Guided Search | [write-up](leaderboard/llvm_instcount/ppo/README.md), [results](leaderboard/llvm_instcount/ppo/results.csv) | 2022-02 | 69.821s  | **1.070×**                   |
| Facebook                                         | Random search (t=10800) | [write-up](leaderboard/llvm_instcount/random_search/README.md), [results](leaderboard/llvm_instcount/random_search/results_p125_t10800.csv) | 2021-03 | 10,512.356s     | 1.062×                       |
| Facebook                                         | Random search (t=3600) | [write-up](leaderboard/llvm_instcount/random_search/README.md), [results](leaderboard/llvm_instcount/random_search/results_p125_t3600.csv) | 2021-03 | 3,630.821s      | 1.061×                       |
| Facebook                                         | Greedy search | [write-up](leaderboard/llvm_instcount/e_greedy/README.md), [results](leaderboard/llvm_instcount/e_greedy/results_e0.csv) | 2021-03 | 169.237s        | 1.055×                       |
| Anthony. W. Jung                                 | GATv2 + DD-PPO | [write-up](leaderboard/llvm_instcount/gatv2_ddppo/README.md), [results](leaderboard/llvm_instcount/gatv2_ddppo/result.csv) | 2022-06 | 258.149s        | 1.047×                       |
| Facebook                                         | Random search (t=60) | [write-up](leaderboard/llvm_instcount/random_search/README.md), [results](leaderboard/llvm_instcount/random_search/results_p125_t60.csv) | 2021-03 | 91.215s         | 1.045×                       |
| Facebook                                         | e-Greedy search (e=0.1) | [write-up](leaderboard/llvm_instcount/e_greedy/README.md), [results](leaderboard/llvm_instcount/e_greedy/results_e10.csv) | 2021-03 | 351.611s        | 1.041×                       |
| Jiadong Guo                                      | Tabular Q (N=5000, H=10) | [write-up](leaderboard/llvm_instcount/tabular_q/README.md), [results](leaderboard/llvm_instcount/tabular_q/results-H10-N5000.csv) | 2021-04 | 2534.305        | 1.036×                       |
| Facebook                                         | Random search (t=10) | [write-up](leaderboard/llvm_instcount/random_search/README.md), [results](leaderboard/llvm_instcount/random_search/results_p125_t10.csv) | 2021-03 | 42.939s         | 1.031×                       |
| Patrick Hesse                                    | DQN (N=4000, H=10) | [write-up](leaderboard/llvm_instcount/dqn/README.md), [results](leaderboard/llvm_instcount/dqn/results-instcountnorm-H10-N4000.csv) | 2021-06 | 91.018s         | 1.029×                       |
| Jiadong Guo                                      | Tabular Q (N=2000, H=5) | [write-up](leaderboard/llvm_instcount/tabular_q/README.md), [results](leaderboard/llvm_instcount/tabular_q/results-H5-N2000.csv) | 2021-04 | 694.105         | 0.988×                       |


## Contributing

We welcome contributions to CompilerGym. If you are interested in contributing please see
[this document](https://github.com/facebookresearch/CompilerGym/blob/development/CONTRIBUTING.md).


## Citation

If you use CompilerGym in any of your work, please cite [our
paper](https://arxiv.org/pdf/2109.08267.pdf):

```
@inproceedings{CompilerGym,
      title={{CompilerGym: Robust, Performant Compiler Optimization Environments for AI Research}},
      author={Chris Cummins and Bram Wasti and Jiadong Guo and Brandon Cui and Jason Ansel and Sahir Gomez and Somya Jain and Jia Liu and Olivier Teytaud and Benoit Steiner and Yuandong Tian and Hugh Leather},
      booktitle={CGO},
      year={2022},
}
```


CHANGELOG.md
## Release 0.2.5 (2022-11-01)

CompilerGym v0.2.5 adds a new LLVM dataset, two new observation spaces, and
includes numerous updates and bug fixes.

**Summary of Changes**

- [llvm] Added two new observation spaces, `LexedIr` and `LexedIrTuple`,
  providing access to a sequence of IR tokens
  ([#742](https://github.com/facebookresearch/CompilerGym/pull/742), thanks
  @fivosts!).
- [llvm] Added the "Jotaibench" benchmark suite, providing 18,761 new executable
  C programs extracted from handwritten code on GitHub
  ([#705](https://github.com/facebookresearch/CompilerGym/pull/705), thanks
  @canesche!).
- Added support for Python 3.10.
- [llvm] Fixed a bug with non-terminating subprocesses
  ([#741](https://github.com/facebookresearch/CompilerGym/pull/741), thanks
  @thecoblack!).
- [llvm] Fixed a bug where the incorrect number of runtimes were reported by
  `reset()` ([#761](https://github.com/facebookresearch/CompilerGym/pull/761)),
  and an incorrect number of warm up runs were being performed
  ([#717](https://github.com/facebookresearch/CompilerGym/pull/717), thanks
  @lqwk!).
- [llvm] New leaderboard submission using GATv2 and DD-PPO
  ([#728](https://github.com/facebookresearch/CompilerGym/pull/728), thanks
  @anthony0727!).
- Added the ability to set timeout on each of the individual environment
  operations ([#716](https://github.com/facebookresearch/CompilerGym/pull/716),
  thanks @ricardoprins!).
* Added support for loading URLs in `CompilerEnvStateReader.read_paths()`
  ([#692](https://github.com/facebookresearch/CompilerGym/pull/692), thanks
  @thecoblack!).
- Simplified Makefile rules: renamed `install-test` to `test` and deprecated
  bazel test rules.
- Fixed a bug where the `TimeLimit` wrapper would interfere with benchmark
  iterator wrappers
  ([#739](https://github.com/facebookresearch/CompilerGym/pull/739), thanks
  @nluu175!).
- [ci] Added CI test coverage of example services
  ([#695](https://github.com/facebookresearch/CompilerGym/pull/695),
  [#642](https://github.com/facebookresearch/CompilerGym/pull/642),
  [#699](https://github.com/facebookresearch/CompilerGym/pull/699), thanks
  @mostafaelhoushi!).
- [ci] Updated Github actions to use Node v16.
- Reduced verbosity and wall time of CMake builds.
- Updates and fixes dependent package conflicts (fixes #771, #768).

**Credits**

A huge thank you to all code contributors!

* @anthony0727
* @canesche made their first contribution in
  [#705](https://github.com/facebookresearch/CompilerGym/pull/705)
* @fivosts made their first contribution in
  [#742](https://github.com/facebookresearch/CompilerGym/pull/742)
* @jaopaulolc made their first contribution in
  [#738](https://github.com/facebookresearch/CompilerGym/pull/738)
* @lqwk made their first contribution in
  [#717](https://github.com/facebookresearch/CompilerGym/pull/717)
* @mostafaelhoushi
* @nluu175 made their first contribution in
  [#739](https://github.com/facebookresearch/CompilerGym/pull/739)
* @ricardoprins made their first contribution in
  [#716](https://github.com/facebookresearch/CompilerGym/pull/716)
* @ryanrussell made their first contribution in
  [#755](https://github.com/facebookresearch/CompilerGym/pull/755)
* @sahirgomez1
* @thecoblack
* @youweiliang made their first contribution in
  [#751](https://github.com/facebookresearch/CompilerGym/pull/751)

**Full Changelog**:
[v0.2.4...v0.2.5](https://github.com/facebookresearch/CompilerGym/compare/v0.2.4...v0.2.5)

## Release 0.2.4 (2022-05-24)

This release adds a new compiler environment, new APIs, and a suite of backend
improvements to improve the flexibility of CompilerGym environments. Many thanks
to code contributors: @sogartar, @KyleHerndon, @SoumyajitKarmakar, @uduse, and
@anthony0727!

Highlights of this release include:

- [mlir] Began work on a new environment for matrix multiplication using MLIR
  ([#652](https://github.com/facebookresearch/CompilerGym/pull/652), thanks
  @KyleHerndon and @sogartar!). Note this environment is not yet included in the
  pypi package and must be [compiled from
  source](https://github.com/facebookresearch/CompilerGym/blob/development/INSTALL.md#building-from-source-with-cmake).
- [llvm] Added a new `env.benchmark_from_clang_invocation()` method
  ([#577](https://github.com/facebookresearch/CompilerGym/pull/577)) that can be
  used for constructing LLVM environment automatically from C/C++ compiler
  invocations. This makes it much easier to integrate CompilerGym with your
  existing build scripts.
- Added three new wrapper classes: `Counter`, that provides op counts for
  analysis ([#683](https://github.com/facebookresearch/CompilerGym/pull/683));
  `SynchronousSqliteLogger`, that provides logging of environment interactions
  to a relational database
  ([#679](https://github.com/facebookresearch/CompilerGym/pull/679)), and
  `ForkOnStep` that provides an `undo()` operation
  ([#682](https://github.com/facebookresearch/CompilerGym/pull/682)).
- Added `reward_space` and `observation_space` parameters to `env.reset()`
  ([#659](https://github.com/facebookresearch/CompilerGym/pull/659), thanks
  @SoumyajitKarmakar!)

This release includes a number of improvements to the backend APIs that make it
easier to write new CompilerGym environments:

- Refactored the backend to make `CompilerEnv` an abstract interface, and
  `ClientServiceCompilerEnv` the concrete implementation of this interface. This
  enables new environments to be implemented without using gRPC
  ([#633](https://github.com/facebookresearch/CompilerGym/pull/633), thanks
  @sogartar!).
- Extended the support for different types of action and observation spaces
  ([#641](https://github.com/facebookresearch/CompilerGym/pull/641),
  [#643](https://github.com/facebookresearch/CompilerGym/pull/643), thanks
  @sogartar!), including new `Permutation` and `SpaceSequence` spaces
  ([#645](https://github.com/facebookresearch/CompilerGym/pull/645), thanks
  @sogartar!)..

www/www.py
# Copyright (c) Facebook, Inc. and its affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
"""A CompilerGym API and web frontend.

This exposes an API with two operations:

   1. /api/v4/describe

        Describe the CompilerGym interface. This generates a list of action
        names and their numeric values, a list of benchmark datasets and the
        benchmarks within them, and a list of reward spaces.

        Example usage:

            $ curl localhost:5000/api/v4/describe
            {
                "actions": {
                    "-adce": 1,
                    ...
                    "-tailcallelim": 122
                },
                "benchmarks": {
                    "benchmark://anghabench-v1": [
                        "8cc/extr_buffer.c_buf_append",
                        ...
                        "8cc/extr_buffer.c_quote_cstring_len"
                    ],
                    "benchmark://blas-v0": [
                        ...
                    ],
                    "benchmark://cbench-v1": [
                        "adpcm",
                        ...
                        "jpeg-c"
                    ],
                    ...
                },
                "rewards": [
                    "IrInstructionCount",
                    ...
                    "ObjectTextSizeOz"
                ]
            }

   2. /ap/v4/step

        Compute the state from the given environment description. Query
        arguments:

            benchmark: The name of the benchmark. If "benchmark_source" is set
                (see below), this is the name of the local file that the user
                selected.

            benchmark_source: An inline string of code to use as the benchmark.

            reward: The name of the reward signal to use.

            actions: An optional, command-separated list of actions to run.

            all_states: An optional string that if "1" means that a list of
                all states will be returned, one for each action. Else, only
                the state for the final action is returned.

        Example usage:

            $ curl 'localhost:5000/api/v4/step?benchmark=benchmark://cbench-v1/adpcm&reward=IrInstructionCountOz&actions=1,2,3'
            {
                "commandline": "opt - ...",
                "rewards": [0.003],
                "done": false,
                "ir": "...",
                "states": [
                    {
                        "instcount": {...},
                        "autophase": {...},
                        "reward": 0.003
                    },
                ]
            }
"""
import logging
import os
import sys
import tempfile
from functools import lru_cache
from itertools import islice
from pathlib import Path
from threading import Lock
from typing import Any, Dict, List, Optional

from flask import Flask, jsonify, request, send_file
from flask_cors import CORS
from pydantic import BaseModel

import compiler_gym
from compiler_gym.datasets.benchmark import Benchmark
from compiler_gym.envs import LlvmEnv
from compiler_gym.envs.llvm import make_benchmark
from compiler_gym.util.truncate import truncate

app = Flask("compiler_gym")
CORS(app)


resource_dir: Path = (Path(__file__).parent / "frontends/compiler_gym/build").absolute()

logger = logging.getLogger(__name__)

# A single compiler environment that is used to serve all endpoints.
env: LlvmEnv = compiler_gym.make("llvm-v0")
env_lock = Lock()


class StateToVisualize(BaseModel):
    """Encapsulates the state to visualize in the frontend."""

    instcount: Dict[str, int]
    autophase: Dict[str, int]
    # The reward signal measures how "good" the previous action was. Over time
    # the sequence of actions that produces the highest cumulative reward is the
    # best:
    reward: float


class StepRequest(BaseModel):
    """User arguments to /api/v4/step."""

    # The name of the benchmark.
    benchmark: str

    # The inline source code for a benchmark.
    benchmark_source: Optional[str]

    # The reward space to use.
    reward: str

    # A comma-separated list of actions to perform.
    actions: List[int]

    # Whether to return a state for every action, or only the final action. See
    # StepReply.states.
    all_states: bool

    @classmethod
    def from_request(cls):
        """Parse the arguments from Flask's request arguments."""

        def required_arg(name: str) -> str:
            value = request.args.get(name)
            if not value:
                raise ValueError(f"Missing requirement argument: {name}")
            return value

        actions_str: str = request.args.get("actions")
        actions: List[int] = (
            [int(x) for x in actions_str.split(",")] if actions_str else []
        )

        return cls(
            benchmark=required_arg("benchmark"),
            benchmark_source=request.args.get("benchmark_source"),
            reward=required_arg("reward"),
            actions=actions,
            all_states=request.args.get("all_states", "0") == "1",
        )


class StepReply(BaseModel):
    """The data returned by a call to /api/v4/step."""

    # This summarizes the sequence of actions that the user has selected so far:
    commandline: str

    # If the compiler environment dies, crashes, or encounters some
    # unrecoverable error, this "done" flag is set. At this point the user
    # should start a new session.
    done: bool

    # The current LLVM-IR:
    ir: str

    # A list of states to visualize, ordered from first to last.
    states: List[StateToVisualize]


@app.route("/api/v4/describe")
def describe():
    with env_lock:
        env.reset()
        return jsonify(
            {
                # A mapping from dataset name to benchmark name. To generate a full
                # benchmark URI, join the two values with a '/'. E.g. given a benchmark
                # "qsort" in the dataset "benchmark://cbench-v1", the full URI is
                # "benchmark://cbench-v1/qsort".
                "benchmarks": {
                    dataset.name: list(
                        islice(
                            (
                                x[len(dataset.name) + 1 :]
                                for x in dataset.benchmark_uris()
                            ),
                            10,
                        )
                    )
                    for dataset in env.datasets
                },
                # A mapping from the name of an action to the numeric value. This
                # numeric value is what is passed as argument to the step() function.
                "actions": {k: v for v, k in enumerate(env.action_space.flags)},
                # A list of reward space names. You select the reward space to use
                # during start().
                "rewards": sorted(list(env.reward.spaces.keys())),
            }
        )


@lru_cache(maxsize=16)
def _make_benchmark(name: str, source: str) -> Benchmark:
    """Construct a benchmark from a file name and contents."""
    with tempfile.TemporaryDirectory() as d:
        tmpfile = Path(d) / Path(name).name
        with open(tmpfile, "w") as f:
            f.write(source)

        try:
            return make_benchmark(tmpfile, timeout=60)
        except Exception as e:
            raise ValueError(f"Failed to compiler benchmark {name}: {e}")


def _step(request: StepRequest) -> StepReply:
    """Run the actual step with parsed arguments."""
    states: List[StateToVisualize] = []

    with env_lock:
        env.reward_space = request.reward

        # Create a benchmark from user-supplied code, or just look up the
        # benchmark by name.
        if request.benchmark_source:
            benchmark = _make_benchmark(request.benchmark, request.benchmark_source)
        else:
            benchmark = request.benchmark

        env.reset(benchmark=benchmark)

        # Replay all actions except the last one.
        if request.all_states:
            # Replay actions one at a time to receive incremental rewards. The
            # first item represents the state prior to any actions.
            (instcount, autophase), _, done, info = env.multistep(
                actions=[],
                observation_spaces=[
                    env.observation.spaces["InstCountDict"],
                    env.observation.spaces["AutophaseDict"],
                ],
            )
            if done:
                raise ValueError(
                    f"Failed to compute initial state: {info['error_details']}"
                )
            states.append(
                StateToVisualize(
                    instcount=instcount,
                    autophase=autophase,
                    reward=0,
                )
            )
            for action in request.actions[:-1]:
                (instcount, autophase), reward, done, info = env.step(
                    action,
                    observation_spaces=[
                        env.observation.spaces["InstCountDict"],
                        env.observation.spaces["AutophaseDict"],
                    ],
                )
                states.append(
                    StateToVisualize(
                        instcount=instcount,
                        autophase=autophase,
                        reward=reward,
                    )
                )
                if done:
                    raise ValueError(
                        f"Failed to apply action {action}: {info['error_details']}"
                    )
        else:
            # Replay actions in a single batch.
            _, _, done, info = env.step(request.actions[:-1])
            if done:
                raise ValueError(
                    f"Failed to apply actions {request.actions}: {info['error_details']}"
                )

        # Perform the final action.
        (ir, instcount, autophase), (reward,), done, _ = env.multistep(
            actions=request.actions[-1:],
            observation_spaces=[
                env.observation.spaces["Ir"],
                env.observation.spaces["InstCountDict"],
                env.observation.spaces["AutophaseDict"],
            ],
            reward_spaces=[env.reward_space],
        )

    states.append(
        StateToVisualize(
            instcount=instcount,
            autophase=autophase,
            reward=reward,
        )
    )
    return StepReply(
        commandline=env.commandline(),
        done=done,
        ir=truncate(ir, max_line_len=250, max_lines=1024),
        states=states,
    )


@app.route("/api/v4/step")
def step() -> Dict[str, Any]:
    try:
        request = StepRequest.from_request()
    except ValueError as e:
        return jsonify({"error": f"Invalid actions: {e}"}), 400

    try:
        return jsonify(_step(request).dict())
    except Exception as e:
        return jsonify({"error": str(e)}), 400


# Web endpoints.


@app.route("/")
def index_resource():
    return send_file(resource_dir / "index.html")


@app.route("/<path>")
def root_resource(path: str):
    return send_file(resource_dir / path)


@app.route("/static/css/<path>")
def css_resource(path: str):
    return send_file(resource_dir / "static/css/" / path)


@app.route("/static/js/<path>")
def js_resource(path: str):
    return send_file(resource_dir / "static/js/" / path)


if __name__ == "__main__":
    logger.setLevel(logging.DEBUG)
    handler = logging.StreamHandler(sys.stderr)
    handler.setLevel(logging.DEBUG)
    formatter = logging.Formatter(
        "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
    )
    handler.setFormatter(formatter)
    logger.addHandler(handler)

    logger.info("Serving from %s", resource_dir)
    app.run(port=int(os.environ.get("PORT", "5000")), host="0.0.0.0")
